src/Crypto/Signing/HsmSignerService.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2020 De Staat der Nederlanden, Ministerie van Volksgezondheid, Welzijn en Sport.
+// Licensed under the EUROPEAN UNION PUBLIC LICENCE v. 1.2
+// SPDX-License-Identifier: EUPL-1.2
+
+using System;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Net.Http.Json;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NL.Rijksoverheid.ExposureNotification.BackEnd.Domain;
+
+namespace NL.Rijksoverheid.ExposureNotification.BackEnd.Crypto.Signing
+{
+    /// <summary>
+    /// Typed HttpClient to request a signature from the HsmSignerService API
+    /// </summary>
+    public class HsmSignerService : IHsmSignerService
+    {
+        private readonly HttpClient _httpClient;
+        private readonly IHsmSignerConfig _config;
+        private readonly ILogger _logger;
+
+        public HsmSignerService(
+            HttpClient httpClient,
+            IHsmSignerConfig config,
+            ILogger<HsmSignerService> logger)
+        {
+            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            _config = config ?? throw new ArgumentNullException(nameof(config));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _httpClient.BaseAddress = new Uri(_config.BaseAddress);
+            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        }
+
+        public async Task<byte[]> GetCmsSignatureAsync(byte[] content)
+        {
+            using var sha256Hasher = SHA256.Create();
+            var contentHash = sha256Hasher.ComputeHash(content);
+            var contentHashBased64 = Convert.ToBase64String(contentHash);
+
+            var cmsPublicCertificateChain = _config.CmsPublicCertificateChain;
+
+            var cmsCertBytes = Encoding.UTF8.GetBytes(cmsPublicCertificateChain);
+            var cmsCertBytesBased64 = Convert.ToBase64String(cmsCertBytes);
+
+            var requestModel = new HsmSignerRequestModel()
+            {
+                Algorithm = "sha256",
+                Cert = cmsCertBytesBased64,
+                Hash = contentHashBased64,
+                TimeStamp = true
+            };
+
+            _httpClient.DefaultRequestHeaders.Remove("Authorization");
+            _httpClient.DefaultRequestHeaders.Add("Authorization", _config.CmsJwt);
+
+            var response = _httpClient.PostAsJsonAsync("/cms", requestModel).Result;
+
+            if (!response.IsSuccessStatusCode)
+            {
+                //TODO: improve error handling
+                _logger.LogError("HTTP request to cms endpoint failed");
+                throw new Exception();
```

Comment:
nit: does it make sense to add the http response code to both the logging and the exception here? Or maybe whatever the response produced it's not successful? When adding it both to the exception and the logging then I'd hope you'd always see the error even if logging is misconfigured / turned off.

Reply:
You're not wrong 😉 I've at least added the status code to the message being logged 👍